### PR TITLE
perf(util/parseHeaders): If the header name is buffer

### LINKED ISSUE #2498: Request to improve performance of stringifying in internal codes
## Bug Description

The cpp binding function, utf8slice called by Buffer.toString() is accounted for a big proportion in the CPU profile resulted from benchmarking a HTTP proxy using undici.

 By using Uint8Arrray, we can eliminat the overhead incurred by the language transition

## Reproducible By

```js
const dispatcher = new Pool('http://127.0.0.1:8090' /* an 4worker nginx */, {
  pipelining: 30,
  connections: 100,
});

const proxy = http.createServer((req, res) => {
  dispatcher.request({ // the constructor of RequestHandler in undici/lib/api/api-request.js
    method: req.method,
    path: req.url,
    headers: req.headers,
    body: req,
  }, (err, { statusCode, headers, body /** a Readable */ }) => {
    if (err) {
      throw err
    }
    res.writeHead(statusCode, headers);
     pipe(body, res);
  });
});

function pipe(src, dst) {
  src.on('readable', () => {
    let chunk;
    while (null !== (chunk = src.read(65536))) {
      dst.write(chunk);
    }
  });
  src.on('end', () => {
    dst.end()
  });
}
```

## Logs & Screenshots

image

## Environment

node v22
linux

### Additional context

I hope the members can provide some guide and more importantly caveats to applying this idea

### PR DESCRIPTION
**Pull request by tsctx**
Fixes #2498

## Benchmark
- *main*
```
• parseHeaders
---------------------------------------------------------------- -----------------------------
parseHeaders               26.12 µs/iter    (21.1 µs … 19.62 ms)   22.5 µs   56.6 µs   72.6 µs
parseHeaders (irregular)   29.83 µs/iter    (25.9 µs … 874.1 µs)   29.3 µs   73.7 µs   80.4 µs

summary for parseHeaders
  parseHeaders
   1.14x faster than parseHeaders (irregular)
```

- *this patch*
```
• parseHeaders
---------------------------------------------------------------- -----------------------------
parseHeaders               14.34 µs/iter    (12.5 µs … 436.2 µs)   13.3 µs   33.5 µs     45 µs
parseHeaders (irregular)   13.99 µs/iter    (12.4 µs … 368.8 µs)   13.1 µs   31.5 µs   39.9 µs

summary for parseHeaders
  parseHeaders (irregular)
   1.03x faster than parseHeaders
```

### CODE REVIEW

File: lib/core/tree.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  /**
+   * @param {Uint8Array} buffer
+   * @returns {string | null}
+   */
+  lookup (buffer) {
+    const length = buffer.length
+    let node = this.#node
+    for (let i = 0; i < length; ++i) {
+      const key = buffer[i]
+      if (!(key in node) || node === undefined) return null
```
Comment (ronag) [inline lib/core/tree.js:35]: Why do you need `key in node`?
```suggestion
      if (node === undefined) return null
```

Comment (tsctx) [reply lib/core/tree.js:35]: In this case, it looks like you don't have to.

File: lib/core/tree.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      node = node[key]
+    }
+    return node === undefined ? null : node.value
```
Comment (ronag) [inline lib/core/tree.js:38]:
```suggestion
    return node === undefined ? this.insert(buffer) : node.value
```

File: lib/core/tree.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    let node = this.#node
+    for (let i = 0; i < target.length; ++i) {
+      const key = target[i]
+      node[key] ??= {}
+      if (key >= 0x61 && key <= 0x7a) {
+        node[key & ~32] ??= node[key]
+      }
+      node = node[key]
+    }
+    node.value = value
```
Comment (ronag) [inline lib/core/tree.js:23]:
```suggestion
    node.value = value
    return value
```

**Comment by metcoder95** [comment]
>Is it safe because it is verified by llhttp?

It should. It might not be a benefit trying to optimize that path

File: lib/core/util.js [hunk 4]
```diff
@@ -219,28 +220,74 @@ function parseKeepAliveTimeout (val) {
...
+    : tree.lookup(value) ?? value.toString().toLowerCase()
+}
+
+/**
+ * @param {Record<any, any> | (Buffer | string | (Buffer | string)[])[]} headers
+ * @param {boolean} [allowUnsafe]
+ * @param {Record<any, any>} [obj]
+ * @returns
+ */
+function parseHeaders (headers, allowUnsafe, obj) {
```
Comment (ronag) [inline lib/core/util.js:246]: Can you move "allowUnsafe" to the end as to not to break the API

Comment (tsctx) [reply lib/core/util.js]: Sure.

**Review by ronag — COMMENTED** [review]
Can you add a benchmark as well?

**Comment by tsctx** [comment]
> Can you add a benchmark as well?

What benchmark library do you use?

**Comment by tsctx** [comment]
> Can you add a benchmark as well?

Which benchmark library should I use?

**Comment by ronag** [comment]
As far as I'm concerned. Whatever you prefer.

**Comment by ronag** [comment]
> @ronag Done.

Are the results in the description the latest ones?

**Review by ronag — APPROVED** [review]

**Comment by tsctx** [comment]
No. I'll update it later.

File: lib/core/tree.js (hunk 1 above)
Comment (ronag) [reply lib/core/tree.js:43]: @tsctx remove it and run benchmark again?

Comment (tsctx) [reply lib/core/tree.js:43]:
```
• parseHeaders
--------------------------------------------------- -----------------------------
default        4.08 µs/iter     (3.3 µs … 905.7 µs)    3.8 µs   10.1 µs   13.2 µs
allowUnsafe    2.38 µs/iter     (2.06 µs … 4.06 µs)   2.62 µs   4.06 µs   4.06 µs

summary for parseHeaders
  allowUnsafe
   1.72x faster than default

• parseHeaders (irregular)
--------------------------------------------------- -----------------------------
default        4.14 µs/iter     (3.88 µs … 5.57 µs)   4.05 µs   5.57 µs   5.57 µs
allowUnsafe    2.24 µs/iter     (2.06 µs … 2.92 µs)   2.25 µs   2.92 µs   2.92 µs

summary for parseHeaders (irregular)
  allowUnsafe
   1.85x faster than default
```

Comment (tsctx) [reply lib/core/tree.js]: It was a little fast.

**Review by ronag — COMMENTED** [review]
Did you try with `Map`?

File: lib/core/tree.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    let node = this.#node
+    for (let i = 0; i < length; ++i) {
+      const key = target[i]
+      node[key] ??= {}
+      if (key >= 0x61 && key <= 0x7a) {
+        node[key & ~32] ??= node[key]
+      }
+      node = node[key]
+    }
+    node.value = value
```
Comment (ronag) [inline lib/core/tree.js:27]: Not sure if it makes a difference, but if the node has only numeric keys that might be faster?

```suggestion
    node[256] = value
```

**Comment by tsctx** [comment]
> @lemire @anonrig Can you think of some other data structure or other way to implement that might be faster? I think the main bottleneck here is the heap-allocated nodes.`Tree`

double array, Ternary Search Tree, LOUDS

**Comment by tsctx** [comment]
It may take some time, but we can implement it.

**Comment by ronag** [comment]
> It may take some time, but we can implement it.

Anything already on npm?

File: lib/core/tree.js (hunk 5 above)
Comment (tsctx) [reply lib/core/tree.js:27]: The benchmark did not change, but I will change it.

**Comment by lemire** [comment]
> Can you think of some other data structure or other way to implement Tree that might be faster? I think the main bottleneck here is the heap-allocated nodes.

I will check this morning.

**Comment by lemire** [comment]
@ronag Give me some time before merging this PR. I am sure it is good but I will do some analysis.

(If only for fun!)

**Comment by anonrig** [comment]
I'll also review it today. Appreciate waiting for a little bit more.

**Comment by lemire** [comment]
If I read this correctly, than this new code has only a chance of being faster if `parseHeaders` is called with a truthy value for `allowUnsafe`, which is not the case currently in this code. That is, no `allowUnsafe` value is passed in the library and so `allowUnsafe` will be falsy. I would suggest turning it on and running realistic benchmarks on undici to determine whether it really does improve the real-world performance issues.

**Comment by lemire** [comment]
The microbenchmark is still worthwhile, but maybe too limited.  See [Benchmarking is hard: processors learn to predict branches](https://lemire.me/blog/2019/10/16/benchmarking-is-hard-processors-learn-to-predict-branches/). I might propose something better.

**Comment by lemire** [comment]
I think that this code is likely quite fast. I think it is correct (though one should always check since small bugs are possible).

Trees are quite fast at classification problems such as those. Their downside is the memory usage. Whether this matters in this case is not something I can tell.

I am not an expert at measuring the memory usage of a data structure in JavaScript. But allow me to use something naive. If I apply the following diff to the PR...

```diff
--- a/lib/core/tree.js
+++ b/lib/core/tree.js
@@ -46,11 +46,11 @@ class Tree {
 }
 
 const tree = new Tree()
-
+console.log(`rss ${Math.round(process.memoryUsage().rss / 1024 / 1024 * 100) / 100} MB`);
 for (let i = 0; i < wellknownHeaderNames.length; ++i) {
   tree.insert(wellknownHeaderNames[i])
 }
-
+console.log(`rss ${Math.round(process.memoryUsage().rss / 1024 / 1024 * 100) / 100} MB`);
 module.exports = {
   Tree,
   tree
```

I get a bump of about 2 MB in the RAM usage with Node.js and Bun. It seems pretty consistent.

The list of allowed `wellknownHeaderNames`, dumped to an ASCII string uses about 1.4 kB. So assuming that the 2 MB result is solid, then the bulk of the added memory usage is due to the tree structure itself.

**Comment by lemire** [comment]
Note that the term `allowUnsafe` seems unwarranted. Or, at least, I don't see why @tsctx's code might be "unsafe". It looks safe, but maybe I missed something.

**Comment by ronag** [comment]
You are right. Unsafe is wrong.

File: lib/core/util.js [hunk 6]
```diff
@@ -219,28 +220,74 @@ function parseKeepAliveTimeout (val) {
...
+ */
+function headerNameToString (value) {
+  const key = value.toString()
+  return headerNameLowerCasedRecord[key] ?? key.toLowerCase()
+}
+
+/**
+ * @param {string | Buffer} value
+ */
+function headerNameToStringUnsafe (value) {
```
Comment (ronag) [inline lib/core/util.js:234]:
```suggestion
function headerNameToString (value) {
```

File: lib/core/util.js [hunk 7]
```diff
@@ -219,28 +220,74 @@ function parseKeepAliveTimeout (val) {
...
+    : tree.lookup(value) ?? value.toString().toLowerCase()
+}
+
+/**
+ * @param {Record<any, any> | (Buffer | string | (Buffer | string)[])[]} headers
+ * @param {Record<any, any>} [obj]
+ * @param {boolean} [allowUnsafe]
+ * @returns
+ */
+function parseHeaders (headers, obj, allowUnsafe) {
```
Comment (ronag) [inline lib/core/util.js:246]:
```suggestion
function parseHeaders (headers, obj) {
```

File: lib/core/util.js [hunk 8]
```diff
@@ -219,28 +220,74 @@ function parseKeepAliveTimeout (val) {
...
+function headerNameToStringUnsafe (value) {
+  return typeof value === 'string'
+    ? headerNameLowerCasedRecord[value] ?? value.toLowerCase()
+    : tree.lookup(value) ?? value.toString().toLowerCase()
+}
+
+/**
+ * @param {Record<any, any> | (Buffer | string | (Buffer | string)[])[]} headers
+ * @param {Record<any, any>} [obj]
+ * @param {boolean} [allowUnsafe]
```
Comment (ronag) [inline lib/core/util.js:243]:
```suggestion
```

File: test/util.js [hunk 9]
```diff
@@ -93,6 +92,17 @@ test('parseHeaders', (t) => {
   t.same(util.parseHeaders([Buffer.from('key'), [Buffer.from('value1'), Buffer.from('value2'), Buffer.from('value3')]]), { key: ['value1', 'value2', 'value3'] })
 })
 
+test('parseHeaders allowUnsafe', (t) => {
```
Comment (ronag) [inline test/util.js:95]:
```suggestion
test('parseHeaders', (t) => {
```

**Comment by lemire** [comment]
@ronag I will have a few more comments *soon*.

File: lib/core/tree.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    if (length > this.#depth) {
+      this.#depth = length
+    }
+  }
+
+  /**
+   * @param {Uint8Array} buffer
+   * @returns {string | null}
+   */
+  lookup (buffer) {
```
Comment (anonrig) [inline lib/core/tree.js:37]: Can you add comments to each of these functions? I'm having hard time following up with the proposed changes.

Comment (tsctx) [reply lib/core/tree.js]: Added.

File: lib/core/tree.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+   * Lowercases key and inserts its value into the node.
+   * @param {string} value
+   */
+  insert (value) {
+    const target = Buffer.from((value = value.toLowerCase()))
+    const length = target.length
+    let node = this.#node
+    for (let i = 0; i < length; ++i) {
+      const key = target[i]
+      node[key] ??= {}
```
Comment (ronag) [inline lib/core/tree.js:17]: Does using an array here make any difference?

Comment (tsctx) [reply lib/core/tree.js:17]: Array space must be allocated.

Comment (tsctx) [reply lib/core/tree.js:17]:
```
node[key] ??= new Array(257)
```

File: benchmarks/parseHeaders.mjs [hunk 12]
```diff
@@ -0,0 +1,52 @@
+import { bench, group, run } from 'mitata'
+import { parseHeaders } from '../lib/core/util.js'
+
+const headers = Object.entries({
```
Comment (lemire) [inline benchmarks/parseHeaders.mjs:4]: I recommend using a more varied input:

```JavaScript

const headers = Array.from([{
  'Content-Type': 'application/json',
  'Date': 'Wed, 01 Nov 2023 00:00:00 GMT',
  'Powered-By': 'NodeJS',
  'Content-Encoding': 'gzip',
  'Set-Cookie': '__Secure-ID=123; Secure; Domain=example.com',
  'Content-Length': '150',
  'Vary': 'Accept-Encoding, Accept, X-Requested-With'
},
{
  "Content-Type": "text/html; charset=UTF-8",
  "Content-Length": "1234",
  "Date": "Wed, 06 Dec 2023 12:47:57 GMT",
  "Server": "Bing"
},
{
  "Content-Type": "image/jpeg",
  "Content-Length": "56789",
  "Date": "Wed, 06 Dec 2023 12:48:12 GMT",
  "Server": "Bing",
  "ETag": "\"a1b2c3d4e5f6g7h8i9j0\""
},
{
  "Cookie": "session_id=1234567890abcdef",
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
  "Host": "www.bing.com",
  "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
  "Accept-Language": "en-US,en;q=0.5",
  "Accept-Encoding": "gzip, deflate, br"
},
{
  "Location": "https://www.bing.com/search?q=bing",
  "Status": "302 Found",
  "Date": "Wed, 06 Dec 2023 12:48:27 GMT",
  "Server": "Bing",
  "Content-Type": "text/html; charset=UTF-8",
  "Content-Length": "0"
},
{
  "Content-Type": "multipart/form-data; boundary=----WebKitFormBoundary1234567890",
  "Content-Length": "98765",
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
  "Host": "www.bing.com",
  "Accept": "*/*",
  "Accept-Language": "en-US,en;q=0.5",
  "Accept-Encoding": "gzip, deflate, br"
},
{
  "Content-Type": "application/json; charset=UTF-8",
  "Content-Length": "2345",
  "Date": "Wed, 06 Dec 2023 12:48:42 GMT",
  "Server": "Bing",
  "Status": "200 OK",
  "Cache-Control": "no-cache, no-store, must-revalidate"
},
{
  "Host": "www.example.com", 
  "Connection": "keep-alive", 
  "Accept": "text/html, application/xhtml+xml, application/xml;q=0.9,;q=0.8", 
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"
}
], (x) => Object.entries(x).flat().map(c => Buffer.from(c)));


const headersIrregular = Object.entries({
  'Content-type': 'application/json',
  'DaTe': 'Wed, 01 Nov 2023 00:00:00 GMT',
  'Powered-by': 'NodeJS',
  'Content-encoding': 'gzip',
  'Set-cookie': '__Secure-ID=123; Secure; Domain=example.com',
  'Content-length': '150',
  'VaRy': 'Accept-Encoding, Accept, X-Requested-With'
}).flat().map(c => Buffer.from(c))
```

File: benchmarks/parseHeaders.mjs [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    parseHeaders(headers, undefined, true)
+  })
+})
+
+group('parseHeaders (irregular)', () => {
+  bench('default', () => {
+    parseHeaders(headersIrregular, undefined, false)
+  })
+  bench('allowUnsafe', () => {
+    parseHeaders(headersIrregular, undefined, true)
```
Comment (lemire) [inline benchmarks/parseHeaders.mjs:46]:
```JavaScript
for(let x of headers) { parseHeaders(headersIrregular, undefined, true); }
```

File: benchmarks/parseHeaders.mjs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+bench('noop', () => {})
+bench('noop', () => {})
+bench('noop', () => {})
+bench('noop', () => {})
+bench('noop', () => {})
+bench('noop', () => {})
+
+group('parseHeaders', () => {
+  bench('default', () => {
+    parseHeaders(headers, undefined, false)
```
Comment (lemire) [inline benchmarks/parseHeaders.mjs:34]:
```JavaScript
for(let x of headers) { fastparseHeaders2(x); }
```

**Comment by lemire** [comment]
@ronag To have some visibility on what is going on, I recommend you try the following patch on the PR:


```diff
--- a/lib/core/tree.js
+++ b/lib/core/tree.js
@@ -9,6 +9,10 @@ class Tree {
     this.#depth = 0
   }
 
+  get_node() {
+    return this.#node
+  }
+
   /**
    * @param {string} value
    */
@@ -46,11 +50,12 @@ class Tree {
 }
 
 const tree = new Tree()
-
 for (let i = 0; i < wellknownHeaderNames.length; ++i) {
   tree.insert(wellknownHeaderNames[i])
 }
 
+console.log(JSON.stringify(tree.get_node(), null, 4))
+
```

I think that it might be important to examine the tree being constructed before merging this so that you are fully aware of what this does. It may not do what one thinks.

I recommend printing it out in full at least once.

**Comment by lemire** [comment]
> Never do JSON.stringfiy because there are multiple references to the same case

Yes, because of the lower/upper case, you have every node twice if you do what I suggest. Thanks for pointing it out. I have updated my proposal: I suggest first pruning the duplicates !!!

**Comment by lemire** [comment]
By pruning the duplicates, you will get a smaller tree. It is still important to look at the constructed data structure as it will reside in memory.

@ronag It is a trade-off. I think @tsctx did a great job speeding up this one function. I trust that it will be 50% in reasonable microbenchmarks. It will also increase the memory usage in a non-trivial manner. I personally do not care because I think we should favour performance over memory usage, and we all have gigabytes of RAM...

**Comment by tsctx** [comment]
I have succeeded in reducing the memory.
Would you prefer this implementation?
It is a little buggy and will take some time.

- before
```
rss 29.63 MB
rss 31.29 MB
```

- after
```
rss 29.45 MB
rss 29.86 MB
```

File: benchmarks/parseHeaders.mjs (hunk 12 above)
Comment (tsctx) [reply benchmarks/parseHeaders.mjs:4]: Thank you.

**Comment by tsctx** [comment]
I have succeeded in reducing the memory.
The trade-off is performance degradation.
Would you prefer this implementation?
It is a little buggy and will take some time.

**Comment by tsctx** [comment]
@ronag @lemire 
I am still trying to reduce the memory as I am a bit concerned about it. As a result I have succeeded in reducing the memory.
The search function has not yet been implemented, but the insert function was.
I think the trade-off is just a small performance loss.
Please let me know if this is a good approach. I would like to work on this continuation.

- before
```
rss 29.71 MB
rss 31.42 MB
```

- after
```
rss 29.71 MB
rss 30.11 MB
```

**Comment by tsctx** [comment]
@ronag @lemire 
I think less memory is better, so we are currently rewriting the tree structure to reduce them.
As a result I have succeeded in reducing the memory.
The search function has not yet been implemented, but the insert function was.
I think the trade-off is just a small performance loss.
Please let me know if this is a good approach. I would like to work on this continuation.

- before
```
rss 29.71 MB
rss 31.42 MB
```

- after
```
rss 29.71 MB
rss 30.11 MB
```

**Comment by tsctx** [comment]
@ronag @lemire 
I think less memory is better, so we are currently rewriting the tree structure to reduce them. As a result I have succeeded in reducing the memory.

The search function has not yet been implemented, but the insert function was.
I think the trade-off is just a small performance loss. Please let me know if this is a good approach. I would like to work on this continuation.

- before
```
rss 29.71 MB
rss 31.42 MB
```

- after
```
rss 29.71 MB
rss 30.11 MB
```

**Comment by ronag** [comment]
TBH I don't care so much for that amount of memory.

**Comment by tsctx** [comment]
It's a little more complicated to implement, but what do you do?

**Comment by ronag** [comment]
SGTM

**Comment by ronag** [comment]
@lemire PTAL

**Comment by tsctx** [comment]
> Can you make a fuzz test for this? i.e. generate lots of random values and compare the results between this and a slower & simpler implementation.

Are you talking about random benchmarks?
Or random tests?

**Comment by tsctx** [comment]
<details>
<summary>Random Benchmark</summary>

```js
import { bench, group, run } from 'mitata'

class Tree {
  #node = {};
  #depth = 0;

  /**
   * @param {Uint8Array} key
   * @param {any} value
   */
  insert(key, value) {
    const length = key.length;
    let node = this.#node;
    for (let i = 0; i < length; ++i) {
      const t = key[i];
      node[t] ??= {};
      // a-z
      if (t >= 0x61 && t <= 0x7a) {
        // Uppercase letters preserve references to lowercase ones.
        node[t & ~32] ??= node[t];
      }
      node = node[t];
    }
    node.value = value;
    if (length > this.#depth) {
      this.#depth = length;
    }
  }

  /**
   * @param {Uint8Array} key
   * @returns {any}
   */
  lookup(key) {
    const length = key.length;
    if (length > this.#depth) return null;
    let node = this.#node;
    for (let i = 0; i < length; ++i) {
      if ((node = node?.[key[i]]) === undefined) return null;
    }
    return node?.value ?? null;
  }
}

class Node {
  /** @type {any} */
  value;
  /** @type {null | Node} */
  left;
  /** @type {null | Node} */
  middle;
  /** @type {null | Node} */
  right;
  /** @type {number} */
  code;
  /**
   * @param {Uint8Array} key
   * @param {any} value
   */
  constructor(key, value) {
    const length = key.length;
    if (length === 0) {
      throw new TypeError("Unreachable");
    }
    this.value = null;
    this.left = null;
    this.middle = null;
    this.right = null;
    this.code = key[0];
    if (length > 1) {
      this.middle = new Node(key.subarray(1), value);
    } else {
      this.value = value;
    }
  }

  /**
   * @param {Uint8Array} key
   * @param {any} value
   */
  add(key, value) {
    const code = key[0];
    if (this.code === code) {
      if (key.length === 1) {
        this.value = value;
      } else if (this.middle !== null) {
        this.middle.add(key.subarray(1), value);
      } else {
        this.middle = new Node(key.subarray(1), value);
      }
    } else if (this.code < code) {
      if (this.left !== null) {
        this.left.add(key, value);
      } else {
        this.left = new Node(key, value);
      }
    } else {
      if (this.right !== null) {
        this.right.add(key, value);
      } else {
        this.right = new Node(key, value);
      }
    }
  }

  /**
   * @param {Uint8Array} key
   * @return {Node | null}
   */
  search(key) {
    const keylength = key.length;
    if (keylength === 0) {
      return null;
    }
    let index = 0;
    let node = this;
    while (node !== null && index < keylength) {
      let code = key[index];
      // A-Z
      if (code >= 0x41 && code <= 0x5a) {
        // Lowercase for uppercase.
        code |= 32;
      }
      while (node !== null) {
        if (code === node.code) {
          if (keylength === ++index) {
            // Returns Node since it is the last key.
            return node;
          }
          node = node.middle;
          break;
        }
        node = node.code < code ? node.left : node.right;
      }
    }
    return null;
  }
}

class TernarySearchTree {
  /** @type {Node | null} */
  node = null;

  /**
   * @param {Uint8Array} key
   * @param {any} value
   * */
  insert(key, value) {
    if (this.node === null) {
      this.node = new Node(key, value);
    } else {
      this.node.add(key, value);
    }
  }

  /**
   * @param {Uint8Array} key
   */
  lookup(key) {
    return this.node?.search(key)?.value ?? null;
  }
}

function generateAsciiString(length) {
  let result = "";
  const characters =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  const charactersLength = characters.length;
  for (let i = 0; i < length; ++i) {
    result += characters[Math.floor(Math.random() * charactersLength)];
  }
  return result;
}

// avoid JIT bias
bench('noop', () => {})
bench('noop', () => {})
bench('noop', () => {})
bench('noop', () => {})
bench('noop', () => {})
bench('noop', () => {})

const tree = new Tree();
const tst  = new TernarySearchTree();

/**@type {string[]} */
const random = new Array(1000).fill(() => generateAsciiString((Math.random() * 100 + 10) | 0)).map((c) => c())
const randomBuffer = random.map(c => Buffer.from(c))

for (let i = 0; i < random.length; ++i) {
  const key = random[i]
  const lowerCasedKey = key.toLowerCase()
  const buffer = Buffer.from(lowerCasedKey)
  tree.insert(buffer, lowerCasedKey)
  tst.insert(buffer, lowerCasedKey)
}

group("all", () => {
  bench("tree", () => {
    for (let i = 0; i < randomBuffer.length; ++i) {
      tree.lookup(randomBuffer[i])
    }
  })
  bench("tst", () => {
    for (let i = 0; i < randomBuffer.length; ++i) {
      tst.lookup(randomBuffer[i])
    }
  }) 
})

await new Promise((resolve) => setTimeout(resolve, 7000))

await run()
```
• all
------------------------------------------------- -----------------------------
tree        16.52 ms/iter    (9.81 ms … 55.34 ms)  16.57 ms  55.34 ms  55.34 ms
tst          3.52 ms/iter     (2.57 ms … 5.75 ms)   3.61 ms      4 ms   5.75 ms

summary for all
  tst
   4.69x faster than tree
```

```

</details>

**Comment by ronag** [comment]
> > Can you make a fuzz test for this? i.e. generate lots of random values and compare the results between this and a slower & simpler implementation.
> 
> Are you talking about random benchmarks? Or random tests?

tests

**Comment by tsctx** [comment]
@ronag 
Is this okey? e3dc3f199533d95c4c27f25fbba2156b8ff17d55

File: test/tree.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
+const { Tree } = require('./utils/tree')
+const { TernarySearchTree } = require('../lib/core/tree')
+const assert = require('assert')
+const { test } = require('tap')
+
+function generateAsciiString (length) {
```
Comment (ronag) [inline test/tree.js:5]: I think we need/can to support any byte character.

File: test/tree.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+  tree.insert(buffer, lowerCasedKey)
+  tst.insert(buffer, lowerCasedKey)
+}
+
+test('all', (t) => {
+  try {
+    for (let i = 0; i < randomBuffer.length; ++i) {
+      const a = tree.lookup(randomBuffer[i])
+      const b = tst.lookup(randomBuffer[i])
+      assert.equal(a, b)
```
Comment (ronag) [inline test/tree.js]:
```suggestion
      assert.equal(a, b.toString().toLowerCase())
```

File: test/tree.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+  const characters =
+    'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
+  const charactersLength = characters.length
+  for (let i = 0; i < length; ++i) {
+    result += characters[Math.floor(Math.random() * charactersLength)]
+  }
+  return result
+}
+
+const tree = new Tree()
```
Comment (ronag) [inline test/tree.js:16]: I think we are killed `Tree` no?

File: test/tree.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+  const lowerCasedKey = key.toLowerCase()
+  const buffer = Buffer.from(lowerCasedKey)
+  tree.insert(buffer, lowerCasedKey)
+  tst.insert(buffer, lowerCasedKey)
+}
+
+test('all', (t) => {
+  try {
+    for (let i = 0; i < randomBuffer.length; ++i) {
+      const a = tree.lookup(randomBuffer[i])
```
[the remark this answers is not in the document]
Comment (tsctx) [reply test/tree.js]: I've already rewritten it this way.
```js
test('all', (t) => {
  for (let i = 0; i < randomBuffer.length; ++i) {
    t.equal(tst.lookup(randomBuffer[i]), tree.lookup(randomBuffer[i]))
  }
  t.end()
})
```

[the remark this answers is not in the document]
Comment (tsctx) [reply test/tree.js]: You mean like this?
```js
test('all', (t) => {
  for (let i = 0; i < randomBuffer.length; ++i) {
    t.equal(tst.lookup(randomBuffer[i]), random[i])
  }
  t.end()
}
```

File: test/tree.js (hunk 15 above)
Comment (tsctx) [reply test/tree.js:5]: I was able to test it correctly using tree.

Comment (ronag) [reply test/tree.js:5]: I don't think tree is a good reference. Why does `toString` fail?

**Review by ronag — APPROVED** [review]

**Review by ronag — APPROVED** [review]

**Comment by lemire** [comment]
I’m planning a blog post on this issue… hopefully today.

**Comment by ronag** [comment]
Regarding memory usage. This will increase minimum usage but also decrease maximum usage as concurrent responses would share memory.

File: lib/core/tree.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
+const { wellknownHeaderNames } = require('./constants')
```
Comment (mcollina) [inline lib/core/tree.js:1]:
```suggestion
'use strict'

const { wellknownHeaderNames } = require('./constants')
```

File: test/tree.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
+const { TernarySearchTree } = require('../lib/core/tree')
```
Comment (mcollina) [inline test/tree.js:1]:
```suggestion
'use strict'

const { TernarySearchTree } = require('../lib/core/tree')
```

File: test/tree.js [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+  .fill(0)
+  .map(() => generateAsciiString((Math.random() * 100 + 5) | 0))
+const randomBuffer = random.map((c) => Buffer.from(c))
+
+for (let i = 0; i < random.length; ++i) {
+  const key = random[i]
+  const lowerCasedKey = key.toLowerCase()
+  const buffer = Buffer.from(lowerCasedKey)
+  tst.insert(buffer, lowerCasedKey)
+}
```
Comment (mcollina) [inline test/tree.js:28]: Please move all the test preparation logic inside the test

**Review by mcollina — CHANGES_REQUESTED** [review]

**Comment by lemire** [comment]
@ronag I finally had time tonight to run some experiments. The blog post is up: [Fast Buffer-to-String conversion in JavaScript with a Lookup Table](https://lemire.me/blog/2023/12/08/fast-buffer-to-string-conversion-in-javascript-with-a-lookup-table/).

**Comment by ronag** [comment]
@lemire I assume you are ok with the current state of this PR then?

**Review by metcoder95 — APPROVED** [review]

**Pull request merged**